src/core/execute.c

```diff
@@ -1079,18 +1079,18 @@ static int enforce_user(const ExecContext *context, uid_t uid) {
                 return 0;
 
         /* Sets (but doesn't look up) the uid and make sure we keep the
-         * capabilities while doing so. */
+         * capabilities while doing so. For setting secure bits the capability CAP_SETPCAP is
+         * required, so we also need keep-caps in this case.
+         */
 
-        if (context->capability_ambient_set != 0) {
+        if (context->capability_ambient_set != 0 || context->secure_bits != 0) {
 
                 /* First step: If we need to keep capabilities but
                  * drop privileges we need to make sure we keep our
                  * caps, while we drop privileges. */
                 if (uid != 0) {
-                        int sb = context->secure_bits | 1<<SECURE_KEEP_CAPS;
-
-                        if (prctl(PR_GET_SECUREBITS) != sb)
-                                if (prctl(PR_SET_SECUREBITS, sb) < 0)
+                        if (!FLAGS_SET(prctl(PR_GET_SECUREBITS), 1<<SECURE_KEEP_CAPS))
+                                if (prctl(PR_SET_SECUREBITS, 1<<SECURE_KEEP_CAPS) < 0)
```

Comment:
hmm, you turn off all securebits we might have already set. is that desirable?

Reply:
i wonder if we should have a helper like this:

```c
int set_securebits(int bits, int mask) {
        int current, applied;
        current = prctl(PR_GET_SECURE_BITS);
        if (current < 0)
                return -errno;
        applied = (current & ~mask)  | (bits & mask);
        if (current == applied)
                return 0;
        if (prctl(PR_SET_SECURE_BITS, applied) < 0)
                return -errno;
        return 1;
}
```

And then we can enable or disable one bit in one call...

Reply:
hmm. I guess at the moment this is no issue.
Reasons:
- enforce_user is only invoked from exec_child
- exec_child is executed directly after forking from init (PID1) and I don't assume the init to have any securebits set
- There are two securebits which affect capabilities during a change of the UID
  - KEEP_CAPS: retains all capabilities in the permitted set, when all zero UIDs are switched to non-zero ones (the capability which shall be set)
  - NO_SETUID_FIXUP: This prevents the kernel to adapt permitted, effective and ambient capability set, when the effective uid is changed between zero and non-zero.
    - Usually the init process has no ambient capabilities set, so only the permitted and effective set are affected by this securebit 
    - The permitted and the effective capabilities are overwritten during execve (for root and non-root)
- after setresuid the securebits are set as configured in the service-file (cmp. line 3746)
- after execve (at the end of exec_child) the capabilities are overwritten

But this might change when enforce_user is used from somewhere else (after setting up the securebits).
So I follow up on your idea, using the helper to add the KEEP_CAPS bit to the securebits. :)